### ISSUE TITLE
Can not reproduce result after setting random seed

### ISSUE THREAD

**Issue by Viscount** [rusty1s/pytorch_geometric#217]
## 🐛 Bug
Pytorch geometric cannot reproduce training result after setting a random seed. We conducted experiments on examples provided by Pytorch-Geometric and got different results on two runs under the same seed.

## To Reproduce

1. add setting seed code in examples/cora.py
`
def set_random_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.backends.cudnn.deterministic = True
    torch.cuda.manual_seed(seed)
    return
`

2. set manual seed before training

`set_random_seed(2333)`

3. run cora.py several times
4. get different accuracy sequence

**The output of 1st run:**

Epoch: 001, Train: 0.4210, Test: 0.4000
Epoch: 002, Train: 0.4385, Test: 0.4040
Epoch: 003, Train: 0.5141, Test: 0.4560
Epoch: 004, Train: 0.6066, Test: 0.5740
Epoch: 005, Train: 0.7242, Test: 0.6840
Epoch: 006, Train: 0.7992, Test: 0.7800
Epoch: 007, Train: 0.8460, Test: 0.8100
Epoch: 008, Train: 0.8724, Test: 0.8200
Epoch: 009, Train: 0.8893, Test: 0.8400
Epoch: 010, Train: 0.9034, Test: 0.8540
.......

**The output of 2nd run:**
Epoch: 001, Train: 0.5445, Test: 0.5200
Epoch: 002, Train: 0.5989, Test: 0.6160
Epoch: 003, Train: 0.5521, Test: 0.5620
Epoch: 004, Train: 0.5785, Test: 0.5780
Epoch: 005, Train: 0.6475, Test: 0.6260
Epoch: 006, Train: 0.7617, Test: 0.6920
Epoch: 007, Train: 0.8583, Test: 0.7820
Epoch: 008, Train: 0.8870, Test: 0.8260
Epoch: 009, Train: 0.8975, Test: 0.8560
Epoch: 010, Train: 0.9052, Test: 0.8680
......

## Expected behavior

The accuracy sequences should be the same

## Environment

* OS: Linux Ubuntu 16.04 LTS
* Python version: 3.6.8
* PyTorch version: 1.0.0
* CUDA/cuDNN version: 9
* GCC version:
* Any other relevant information:

**Comment by Viscount** [comment]
> ## 🐛 Bug
> Pytorch geometric cannot reproduce training result after setting a random seed. We conducted experiments on examples provided by Pytorch-Geometric and got different results on two runs under the same seed.
> 
> ## To Reproduce
> Steps to reproduce the behavior:
> 
> 1. add setting seed code in examples/cora.py
>    `def set_random_seed(seed): random.seed(seed) np.random.seed(seed) torch.manual_seed(seed) torch.backends.cudnn.deterministic = True torch.cuda.manual_seed(seed) return`
> 2. set manual seed before training
> 
> `set_random_seed(2333)`
> 
> 1. run cora.py several times
> 2. get different accuracy sequence
> 
> **The output of 1st run:**
> 
> Epoch: 001, Train: 0.4210, Test: 0.4000
> Epoch: 002, Train: 0.4385, Test: 0.4040
> Epoch: 003, Train: 0.5141, Test: 0.4560
> Epoch: 004, Train: 0.6066, Test: 0.5740
> Epoch: 005, Train: 0.7242, Test: 0.6840
> Epoch: 006, Train: 0.7992, Test: 0.7800
> Epoch: 007, Train: 0.8460, Test: 0.8100
> Epoch: 008, Train: 0.8724, Test: 0.8200
> Epoch: 009, Train: 0.8893, Test: 0.8400
> Epoch: 010, Train: 0.9034, Test: 0.8540
> .......
> 
> **The output of 2nd run:**
> Epoch: 001, Train: 0.5445, Test: 0.5200
> Epoch: 002, Train: 0.5989, Test: 0.6160
> Epoch: 003, Train: 0.5521, Test: 0.5620
> Epoch: 004, Train: 0.5785, Test: 0.5780
> Epoch: 005, Train: 0.6475, Test: 0.6260
> Epoch: 006, Train: 0.7617, Test: 0.6920
> Epoch: 007, Train: 0.8583, Test: 0.7820
> Epoch: 008, Train: 0.8870, Test: 0.8260
> Epoch: 009, Train: 0.8975, Test: 0.8560
> Epoch: 010, Train: 0.9052, Test: 0.8680
> ......
> 
> ## Expected behavior
> The accuracy sequences should be the same
> 
> ## Environment
> * OS: Linux Ubuntu 16.04 LTS
> * Python version: 3.6.8
> * PyTorch version: 1.0.0
> * CUDA/cuDNN version: 9
> * GCC version:
> * Any other relevant information:
> 
> ## Additional context

**Comment by rusty1s** [comment]
Can you verify that it works on CPU?

**Comment by Viscount** [comment]
> Can you verify that it works on CPU?

Seems it doesn't work on CPU.
`device = torch.device('cpu')` 
After setting the device to CPU only, I still got different result sequences on the same seed. Here's the result on my device. I've attached the code file in TXT format.
[cora.txt](https://github.com/rusty1s/pytorch_geometric/files/3098009/cora.txt)

Run #1:
Epoch: 001, Train: 0.3015, Test: 0.3380
Epoch: 002, Train: 0.3367, Test: 0.3700
Epoch: 003, Train: 0.4028, Test: 0.4080
Epoch: 004, Train: 0.5597, Test: 0.5260
Epoch: 005, Train: 0.7412, Test: 0.7060
...
Run #2: 
Epoch: 001, Train: 0.3396, Test: 0.3580
Epoch: 002, Train: 0.4760, Test: 0.4700
Epoch: 003, Train: 0.6153, Test: 0.5740
Epoch: 004, Train: 0.7090, Test: 0.6620
Epoch: 005, Train: 0.7799, Test: 0.7300
...

**Comment by rusty1s** [comment]
Hi, this is actually not reproducible for me. A simple `torch.manual_seed(12345)` after all imports is enough to make both `gcn.py` and `cora.py` deterministic (and `SplineConv` runtime could certainly be improved on CPU).

Run 1:
```
Epoch: 001, Train: 0.3583, Test: 0.3240
Epoch: 002, Train: 0.5907, Test: 0.5380
```
Run 2:
```
Epoch: 001, Train: 0.3583, Test: 0.3240
Epoch: 002, Train: 0.5907, Test: 0.5380
```

**Comment by Viscount** [comment]
THX, I'll try this seed on my device and report results.

**Comment by Viscount** [comment]
> Hi, this is actually not reproducible for me. A simple `torch.manual_seed(12345)` after all imports is enough to make both `gcn.py` and `cora.py` deterministic (and `SplineConv` runtime could certainly be improved on CPU).
> 
> Run 1:
> 
> ```
> Epoch: 001, Train: 0.3583, Test: 0.3240
> Epoch: 002, Train: 0.5907, Test: 0.5380
> ```
> Run 2:
> 
> ```
> Epoch: 001, Train: 0.3583, Test: 0.3240
> Epoch: 002, Train: 0.5907, Test: 0.5380
> ```

Got the same results as yours. It seems that I put the seed-setting code in the wrong place. This issue can be closed.

**Comment by yonatansverdlov** [comment]
> > Hi, this is actually not reproducible for me. A simple `torch.manual_seed(12345)` after all imports is enough to make both `gcn.py` and `cora.py` deterministic (and `SplineConv` runtime could certainly be improved on CPU).
> > Run 1:
> > ```
> > Epoch: 001, Train: 0.3583, Test: 0.3240
> > Epoch: 002, Train: 0.5907, Test: 0.5380
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Run 2:
> > ```
> > Epoch: 001, Train: 0.3583, Test: 0.3240
> > Epoch: 002, Train: 0.5907, Test: 0.5380
> > ```
> 
> Got the same results as yours. It seems that I put the seed-setting code in the wrong place. This issue can be closed.

Where did you set your seed ?

**Comment by yonatansverdlov** [comment]
Where did you put the seed ?

**Comment by rusty1s** [comment]
I usually set it right at the beginning of a script, e.g.:
```
from torch_geometric import seed_everything

seed_everything(12345)
